rustls/src/server/tls13.rs

```diff
@@ -581,40 +581,39 @@ impl CompleteClientHelloHandling {
             .collect();
 
         let supported_groups = suites::KeyExchange::supported_groups();
-        let chosen_group = supported_groups
+
+        let chosen_share = supported_groups
             .iter()
-            .find(|group| share_groups.contains(group))
-            .cloned();
+            .find(|group| share_groups.contains(group)) // Find the chosen group
+            .and_then(|g| shares_ext.iter().find(|k| k.group == *g)); // Then its share
 
-        if chosen_group.is_none() {
-            // We don't have a suitable key share.  Choose a suitable group and
-            // send a HelloRetryRequest.
-            let retry_group_maybe = supported_groups
+        let chosen_share =  match chosen_share {
+            Some(s) => s,
+            None => {
+                // We don't have a suitable key share.  Choose a suitable group and
+                // send a HelloRetryRequest.
+                let retry_group_maybe = supported_groups
                 .iter()
                 .find(|group| groups_ext.contains(group))
                 .cloned();
-            self.handshake
-                .transcript
-                .add_message(chm);
 
-            if let Some(group) = retry_group_maybe {
-                if self.done_retry {
-                    return Err(hs::illegal_param(sess, "did not follow retry request"));
-                }
+                self.handshake
+                    .transcript
+                    .add_message(chm);
 
-                self.emit_hello_retry_request(sess, group);
-                self.emit_fake_ccs(sess);
-                return Ok(self.into_expect_retried_client_hello());
-            }
+                if let Some(group) = retry_group_maybe {
+                    if self.done_retry {
```

Comment:
> It would be clearer to check self.done_retry error out before we try to calculate retry_group_maybe.

We should probably do that reorganization to make it clearer, but my suggestion is out of scope for this PR.